PublicMapping/districtbuilder
# Add HAR files & load test harness to replicate healthcheck timeouts

## Overview

We had an existing `k6` based load-testing suite, but it was outdated and needed to be updated, so I replaced what was there with a harness script similar to what we've used in other projects (OAR, Raster Foundry, etc.), capable of loading a .HAR file with any set of archived requests (generated from the network tab of the browser developer tools).

The .HAR files I uploaded use a local developer environment, but the domain is overridable via an environment variable (overiding like this is required to use `k6` through Docker regardless).

### Checklist


### Demo

![image](https://user-images.githubusercontent.com/4432106/138470551-312197a4-4b13-46c6-95b3-b3382be04cfb.png)


### Notes

 - I spent some time trying to support testing both the healthcheck & project endpoints from 1 script, but in the end found it easier to run two instances of `k6` in separate tabs, one running the health check .HAR & the other running the project editing .HAR file
 - There is support in `k6` for creating virtual users by duplicating requests, but I couldn't use this for the project `PATCH` requests, as we really wanted to simulate concurrent access to separate projects, which is why I re-invented a tiny variant of this virtual user support myself.

## Testing Instructions
- Ensure you have a PA project setup and note the UUID (use this for `PROJECT_UUIDS` in the example below)
- Open the network tab in the developer tools and find the `Authorization` header for an API request, and note the value (use the value after `Bearer ` in the examples below)
- Using the directions in the `load-testing/README.md`, in three separate shells run the following:
  - `scripts/server` (make sure to wait for all topology layers to finish loading)
  - `JWT_AUTH_TOKEN="..." HAR_FILE=patch-local-pa-project.har PROJECT_UUIDS="..." REQ_ORIGIN="http://host.docker.internal:3003" docker-compose run --rm k6`
   - `JWT_AUTH_TOKEN="..." HAR_FILE=healthcheck.har REQ_ORIGIN="http://host.docker.internal:3005" docker-compose run --rm k6 run /scripts/harness.js --iterations=4`
- You should see the above load tests both succeed
- Duplicate your PA project a number of times, and include those additional UUIDS as part of `PROJECT_UUIDS`, and rerun the load testing scripts. You should be able to reliably cause the healthchecks to fail under sufficient load (10 concurrent projects being edited was enough for me)

Closes #1015
